psi4/src/psi4/libsapt_solver/usapt0.cc

```diff
@@ -0,0 +1,2535 @@
+            r2B += C_DDOT(no*nv,rp[0],1,rp[0],1);
+            r2B = sqrt(r2B) / b2B;
+        }
+
+        stop = time(NULL);
+        outfile->Printf( "    %-4d %11.3E%1s %11.3E%1s %10ld\n", iter+1,
+            r2A, (r2A < delta_ ? "*" : " "),
+            r2B, (r2B < delta_ ? "*" : " "),
+            stop-start
+            );
+
+        if (r2A <= delta_ && r2B <= delta_) {
+            break;
+        }
+
+        if (r2A > delta_) {
+            preconditioner(ra_A,za_A,eps_occa_A_,eps_vira_A_);
+            preconditioner(rb_A,zb_A,eps_occb_A_,eps_virb_A_);
+            double zr_new = za_A->vector_dot(ra_A);
+            zr_new += zb_A->vector_dot(rb_A);
+            double beta = zr_new / zr_old_A;
+            zr_old_A = zr_new;
+
+            int no = pa_A->nrow();
+            int nv = pa_A->ncol();
+            double** pp = pa_A->pointer();
+            double** zp = za_A->pointer();
+            C_DSCAL(no*nv,beta,pp[0],1);
+            C_DAXPY(no*nv,1.0,zp[0],1,pp[0],1);
+            no = pb_A->nrow();
+            nv = pb_A->ncol();
+            pp = pb_A->pointer();
+            zp = zb_A->pointer();
+            C_DSCAL(no*nv,beta,pp[0],1);
+            C_DAXPY(no*nv,1.0,zp[0],1,pp[0],1);
+        }
+
+        if (r2B > delta_) {
+            preconditioner(ra_B,za_B,eps_occa_B_,eps_vira_B_);
+            preconditioner(rb_B,zb_B,eps_occb_B_,eps_virb_B_);
+            double zr_new = za_B->vector_dot(ra_B);
+            zr_new += zb_B->vector_dot(rb_B);
+            double beta = zr_new / zr_old_B;
+            zr_old_B = zr_new;
+
+            int no = pa_B->nrow();
+            int nv = pa_B->ncol();
+            double** pp = pa_B->pointer();
+            double** zp = za_B->pointer();
+            C_DSCAL(no*nv,beta,pp[0],1);
+            C_DAXPY(no*nv,1.0,zp[0],1,pp[0],1);
+            no = pb_B->nrow();
+            nv = pb_B->ncol();
+            pp = pb_B->pointer();
+            zp = zb_B->pointer();
+            C_DSCAL(no*nv,beta,pp[0],1);
+            C_DAXPY(no*nv,1.0,zp[0],1,pp[0],1);
+        }
+    }
+
+    outfile->Printf( "    -----------------------------------------\n");
+    outfile->Printf( "\n");
+
+    if (iter == maxiter_)
+        throw PSIEXCEPTION("CPKS did not converge.");
+}
+void CPKS_USAPT0::preconditioner(std::shared_ptr<Matrix> r,
+                               std::shared_ptr<Matrix> z,
+                               std::shared_ptr<Vector> o,
+                               std::shared_ptr<Vector> v)
+{
+    int no = o->dim();
+    int nv = v->dim();
+
+    double** rp = r->pointer();
+    double** zp = z->pointer();
+
+    double* op = o->pointer();
+    double* vp = v->pointer();
+
+    for (int i = 0; i < no; i++) {
+        for (int a = 0; a < nv; a++) {
+            zp[i][a] = rp[i][a] / (vp[a] - op[i]);
+        }
+    }
+}
+
+std::map<std::string, std::shared_ptr<Matrix> > CPKS_USAPT0::product(std::map<std::string, std::shared_ptr<Matrix> >& b)
+{
```

Comment:
All of the wavefunctions have Hessian Vector products on them now, see [here](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libscf_solver/uhf.cc#L372). Its not quite the same thing as you want to compute A and B monomers at the same time, but it would be nice if we could consolidate this interface so we stop rewriting these in many places. Whats nice about the Hx products on the respective wavefunctions is they will be updated to include DFT components as well later this year. 

This could also help with your CPROHF issue, the ROHF Hx products are already baked in so coupled induction should be pretty easy.

Let me know if you see any way to combine these.